lib/Transforms/LazyRelin/LazyRelin.cpp

```diff
@@ -0,0 +1,220 @@
+#include "lib/Transforms/LazyRelin/LazyRelin.h"
+
+#include <numeric>
+
+#include "lib/Analysis/LazyRelinAnalysis/LazyRelinAnalysis.h"
+#include "lib/Dialect/BGV/IR/BGVDialect.h"
+#include "lib/Dialect/BGV/IR/BGVOps.h"
+#include "lib/Dialect/LWE/IR/LWETypes.h"
+#include "lib/Dialect/Utils.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"           // from @llvm-project
+#include "llvm/include/llvm/Support/Debug.h"            // from @llvm-project
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/IRMapping.h"             // from @llvm-project
+#include "mlir/include/mlir/IR/Visitors.h"              // from @llvm-project
+#include "mlir/include/mlir/Pass/Pass.h"                // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+#define DEBUG_TYPE "LazyRelin"
+
+#define GEN_PASS_DEF_LAZYRELIN
+#include "lib/Transforms/LazyRelin/LazyRelin.h.inc"
+
+struct LazyRelin : impl::LazyRelinBase<LazyRelin> {
+  using LazyRelinBase::LazyRelinBase;
+
+  void processFunc(func::FuncOp funcOp) {
+    MLIRContext *context = funcOp->getContext();
+
+    // Remove all relin ops. This makes the IR invalid, because the key basis
+    // sizes are incorrect. However, the correctness of the ILP ensures the key
+    // basis sizes are made correct at the end.
+    funcOp->walk([&](bgv::RelinearizeOp op) {
+      op.getResult().replaceAllUsesWith(op.getOperand());
+      op.erase();
+    });
+
+    LazyRelinAnalysis analysis(funcOp);
+    if (failed(analysis.solve())) {
+      funcOp->emitError("Failed to solve the optimization problem");
+      return signalPassFailure();
+    }
+
+    OpBuilder b(&getContext());
+
+    funcOp->walk([&](Operation *op) {
+      if (!analysis.shouldInsertRelin(op)) return;
+
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Inserting relin after: " << op->getName() << "\n");
+
+      b.setInsertionPointAfter(op);
+      for (Value result : op->getResults()) {
+        // Fill a vector with 0, 1, ..., degree
+        int fromBasisDegree = 1 + analysis.keyBasisDegreeBeforeRelin(result);
+        std::vector<int> range(fromBasisDegree);
+        std::iota(std::begin(range), std::end(range), 0);
+
+        DenseI32ArrayAttr beforeBasis = b.getDenseI32ArrayAttr(range);
+        DenseI32ArrayAttr afterBasis = b.getDenseI32ArrayAttr({0, 1});
+        auto reduceOp = b.create<bgv::RelinearizeOp>(op->getLoc(), result,
+                                                     beforeBasis, afterBasis);
+        result.replaceAllUsesExcept(reduceOp.getResult(), {reduceOp});
+      }
+    });
+
+    // At this point we need to do some cleanup. The process of removing the
+    // initial relinearize ops and inserting new ones did not update the key
+    // basis sizes on result types. So we walk the IR and update them if
+    // necessary.
+    funcOp->walk([&](Operation *op) {
```

Comment:
This seems complicated enough that it should maybe be a separate utils thing to massage types back into consistency?

Reply:
The most complicated part are all the guards, so I moved that into a local helper function in 8e657c05. I don't know how much of the result type patching can be reused, because the generic part is the operation cloner/replacer. Maybe we could do connect it to type inference hooks somehow, like "re-infer the result types and replace the op with its correct result type", but that is beyond the scope of this PR.

Reply:
Thanks, that's a great first step! I agree my full request would be too much work for this PR